File: lab3_16_12/2.cpp
#include<bits/stdc++.h>

using namespace std;

void printPath(vector<int>& parent_src, vector<int>& parent_dest, int m, int src, int dest)
{
    vector<int> path;

    int curr = m;
    path.push_back(curr);

    while(curr != src)
    {
        curr = parent_src[curr];
        path.push_back(curr);
    }

    reverse(path.begin(), path.end());

    curr = m;
    while(curr != dest)
    {
        curr = parent_dest[curr];
        path.push_back(curr);
    }

    cout << "Path found: ";
    for(int i : path)
        cout << i << " ";
    cout << "\n";
}

void biDFS(vector<vector<int>>& adj, int src, int dest, int n)
{
    if(src == dest)
    {
        cout << "src & dest are same.\n";
        return;
    }

    stack<int> s_src, s_dest;

    vector<int> vis_src(n, 0);
    vector<int> vis_dest(n, 0);

    vector<int> parent_src(n, -1);
    vector<int> parent_dest(n, -1);

    s_src.push(src);
    s_dest.push(dest);

    vis_src[src] = 1;
    vis_dest[dest] = 1;

    while(!s_src.empty() && !s_dest.empty())
    {
        // Forward DFS Step
        if(!s_src.empty())
        {
            int u = s_src.top();
            s_src.pop();

            for(int v = 0; v < n; v++)
            {
                if(adj[u][v] == 1 && !vis_src[v])
                {
                    vis_src[v] = 1;
                    parent_src[v] = u;
                    s_src.push(v);

                    if(vis_dest[v])
                    {
                        printPath(parent_src, parent_dest, v, src, dest);
                        return;
                    }
                }
            }
        }

        //  Backward DFS Step
        if(!s_dest.empty())
        {
            int u = s_dest.top();
            s_dest.pop();

            for(int v = 0; v < n; v++)
            {
                if(adj[u][v] == 1 && !vis_dest[v])
                {
                    vis_dest[v] = 1;
                    parent_dest[v] = u;
                    s_dest.push(v);

                    if(vis_src[v])
                    {
                        printPath(parent_src, parent_dest, v, src, dest);
                        return;
                    }
                }
            }
        }
    }

    cout << "No path exists\n";
}

int main()
{
    int n;
    cout << "Enter number of nodes: ";
    cin >> n;

    vector<vector<int>> adj(n, vector<int>(n, 0));

    cout << "Enter Adjacency Matrix (" << n << "x" << n << "):\n";
    for(int i = 0; i < n; i++)
    {
        for(int j = 0; j < n; j++)
        {
            cin >> adj[i][j];
        }
    }

    int src, dest;
    cout << "Enter Source and Destination: ";
    cin >> src >> dest;

    biDFS(adj, src, dest, n);

    return 0;
}

File: lab3_16_12/2.py
import sys

def print_path(parent_src, parent_dest, m, src, dest):
    path_src = []
    curr = m
    while curr != -1:
        path_src.append(curr)
        if curr == src: break
        curr = parent_src[curr]
    path_src.reverse()

    path_dest = []
    curr = parent_dest[m]
    while curr != -1:
        path_dest.append(curr)
        if curr == dest: break
        curr = parent_dest[curr]
    
    full_path = path_src + path_dest
    print("Path found:", *(full_path))

def bidirectional_dfs(adj, src, dest, n):
    if src == dest:
        print("src & dest are same.")
        return

    s_src = [src]
    s_dest = [dest]

    vis_src = [0] * n
    vis_dest = [0] * n
    parent_src = [-1] * n
    parent_dest = [-1] * n

    vis_src[src] = 1
    vis_dest[dest] = 1

    while s_src and s_dest:
        # Forward DFS Step
        if s_src:
            u = s_src.pop()
            for v in range(n):
                if adj[u][v] == 1 and not vis_src[v]:
                    vis_src[v] = 1
                    parent_src[v] = u
                    s_src.append(v)
                    if vis_dest[v]:
                        print_path(parent_src, parent_dest, v, src, dest)
                        return

        # Backward DFS Step
        if s_dest:
            u = s_dest.pop()
            for v in range(n):
                if adj[u][v] == 1 and not vis_dest[v]:
                    vis_dest[v] = 1
                    parent_dest[v] = u
                    s_dest.append(v)
                    if vis_src[v]:
                        print_path(parent_src, parent_dest, v, src, dest)
                        return

    print("No path exists")

def main():
    try:
        input_data = sys.stdin.read().split()
        if not input_data: return
        it = iter(input_data)
        
        n = int(next(it))
        adj = []
        for i in range(n):
            row = []
            for j in range(n):
                row.append(int(next(it)))
            adj.append(row)

        src = int(next(it))
        dest = int(next(it))

        bidirectional_dfs(adj, src, dest, n)
    except (StopIteration, ValueError, EOFError):
        pass

if __name__ == "__main__":
    main()
